lib/layer/layer.dart

```diff
@@ -0,0 +1,71 @@
+import 'package:meta/meta.dart';
+import 'dart:ui';
+
+import './processors.dart';
+export './processors.dart';
+
+abstract class Layer {
+  List<LayerProcessor> preProcessors = [];
+  List<LayerProcessor> postProcessors = [];
+
+  Picture _picture;
+
+  PictureRecorder _recorder;
+  Canvas _canvas;
+
+  @mustCallSuper
+  void render(Canvas canvas, {double x = 0.0, double y = 0.0}) {
+    if (_picture == null) {
+      return;
+    }
+
+    canvas.save();
+    canvas.translate(x, y);
+
+    preProcessors.forEach((p) => p.process(_picture, canvas));
+    canvas.drawPicture(_picture);
+    postProcessors.forEach((p) => p.process(_picture, canvas));
+    canvas.restore();
+  }
+
+  Canvas get canvas {
+    assert(
+      _canvas != null,
+      'Layer is not ready for rendering, call beginRendering first',
+    );
+    return _canvas;
+  }
+
+  void beginRendering() {
+    _recorder = PictureRecorder();
+    _canvas = Canvas(_recorder);
+  }
+
+  void finishRendering() {
+    _picture = _recorder.endRecording();
+
+    _recorder = null;
+    _canvas = null;
+  }
+
+  void drawLayer();
+}
+
+abstract class PreRenderedLayer extends Layer {
+  PreRenderedLayer() {
+    beginRendering();
+    drawLayer();
+    finishRendering();
+  }
```

Comment:
how do I trigger a re-render if something changes? Like I envision myself using this for a tilemap for example. But sometimes some action must trigger a full re-render. How about extracting `begin(); draw() finish()` to a method reRender() and then you can use it here, on the render method of dynamic & the user can call it on a PreRenderedLayer to force an update?

Reply:
Okay, gonna make the reRender method